Jit fuse clamp

t-vi opened the pull request:
This patch adds fused forward and backward for clamp to the jit.
This is one item of #11118 . If it's OK, I'd be happy to also add some more of #11118 .

The patch depends on #11150 , which I merged into master as a base. I'll rebase it when that or #10981 is merged.

This is first serious jit patch, thank you, @ngimel and the others for their guidance. All errors are my own.

Commit:
return aten::gt to the list of fusable operations, add expected graphs

Commit:
return aten::gt to the list of fusable operations, add expected graphs

Commit:
lint

Commit:
lint

Commit:
modify tests to check that all nodes are fused

t-vi committed:
Merge branch 'pr_11150' into jit_fuse_clamp

t-vi committed:
fuse clamp forward and backward

ngimel commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -481,6 +481,8 @@ std::string encodeRHS(Node * n) {
     {aten::sub, "(${0} - ${2}*${1})"},
     {aten::rand_like, "uniform(rnd())"},
+    //min, max
+    {aten::clamp, "fmaxf(fminf(${0}, ${2}), ${1})"},
```
I did not look up how clamp is implemented in the core, but this won't propagate nans (fmaxf and fminf return other when comparing nan with other according to standard) and we usually try to propagate nans. A few lines above, aten::relu is implemented as `{aten::relu, "${0} < 0 ? 0.f : ${0} "},` for this reason.

apaszke replied:
Well we don't need to maintain the same exact NaN semantics in the JIT I think (it's kind of UB for most of our ops), but we don't need to use the intrinsics, because they might end up compiling to code slower than one that uses simple comparisons.

t-vi replied:
While there might be code not providing proper NaN semantics, I'm kind of attached to them.
For clamp, we actually have min/max=NaN to mean "unbounded", so `(a > max ? max : a)` gives the right thing.

apaszke commented:
Unfortunately this will conflict with #10981 😕

t-vi commented:
I thought the plan was to have #10981 go first, then #11150 and then this...
Gives me some time to look at the other ones, at least threshold and fix the above.

t-vi committed:
handle NaN

ngimel commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -119,9 +119,13 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+      // boundary and the factor is 1 when the boundary is NaN
+      // the ! is expressed as "1-" for lack of a "not" function and
+      // the the fuser insisting on float
+      return {(inputs.at(0).isnan() ? inputs.at(0) : grads.at(0))
```
I don't think it will work, it's not an elementwise traversal. isnan() returns tensor which will always evaluate to true, so you'll always be getting inputs.at(0) here.

t-vi replied:
@ngimel Thanks for pointing that out! The sweet poison of overloaded symbolic operators made me think that anything would work...
@vishwakftw I'm trying to stick with fusable operators. I think at::where might count eventually, but only when you can have byte inputs which I understand you cannot at the moment.

ngimel replied:
You can have byte inputs, but you'd have to add some special casing in the autodiff pass and fusion pass.

t-vi replied:
Fusing where might be attractive, then...

t-vi committed:
Fix clamp gradient

Commit:
modify tests to check that all nodes are fused

Commit:
lint

t-vi committed:
Merge branch 'pr_11150' into jit_fuse_clamp

apaszke commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -106,6 +107,19 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+      // but that is hard to reliably code here, so we have 0 as gradient
+      // when the input is NaN (unless grads is NaN or infinite)
+      return {grads.at(0)
+	      * (1-(inputs.at(0).isnan()).type_as(inputs.at(0)))
```
This term is missing from the derivative formula we usually use for clamp. I'm not sure why do we want it. I don't think it's worth sacrifising performance just to change the NaN handling.

t-vi replied:
Personally, I think that proper NaN handling is a good feature and was about to fix it, but I'll take it out if the policy is not to treat NaN. (I do seem to recall from the forum that it caused quite bit of confusion that ReLU forward used to swallow NaN.)

apaszke reviewed (approved):
LGTM, but I'd like to simplify the derivative formula

ezyang committed:
Merge branch 'master' into jit_fuse_clamp

t-vi commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -106,6 +107,19 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
       return {grads.at(0) * (outputs.at(0) > at::Scalar(0)).type_as(outputs.at(0))};
 
+    } else if (node->matches("aten::clamp(Tensor self, Scalar min, Scalar max) -> Tensor")) {
+      // we do two type_as as it's free (hopefully) and the "*" only works with float
```
Thanks, I'll not claim that.

t-vi committed:
Address review comments

Thank you, @apaszke, for your feedback!

Also put back support for DifferentiableGraph in assertAllFused.

t-vi commented:
@ezyang I think it is good, but I don't know what to make out of the CI failure - it doesn't look related at first sight but from clicking on "previous build" a couple of times, it's apparently not shared by other PRs...
